src/lua/protobuf.lua

```diff
@@ -0,0 +1,652 @@
+    end
+    if value_by_id[0] == nil then
+        error(('%q definition does not contain a field with id = 0'):
+            format(enum_name))
+    end
+    return {
+        type = 'enum',
+        name = enum_name,
+        id_by_value = id_by_value,
+        value_by_id = value_by_id,
+    }
+end
+
+-- Create a protocol object that stores message
+-- data needed for encoding.
+--
+-- Accepts protocol definition using protobuf.message
+-- and protobuf.enum functions as in example.
+--
+-- protocol_def = {
+--     protocol.message(<message_name>, <message_def>),
+--     protocol.enum(<enumc_name>, <enum_def>),
+--     <...>
+-- }
+local function protocol(protocol_def)
+    local return_protocol = {}
+    -- Declaration table is used to check forward declarations
+    -- false -- this type used as the field type in the message was not defined
+    -- true -- this field type was defined
+    local declarations = {}
+    for _, def in pairs(protocol_def) do
+        if declarations[def.name] then
+            error(('Double definition of name %q'):format(def.name))
+        end
+        if def.type == 'message' then
+            for _, field_def in pairs(def.field_by_id) do
+                local standard = scalars[field_def.type] ~= nil
+                local declared = declarations[field_def.type]
+                if field_def.type == def.name then
+                    error(('Message %q has a field of %q type ' ..
+                        'Recursive definition is not allowed'):format(
+                        def.name, field_def.type))
+                elseif not standard and not declared then
+                    declarations[field_def.type] = false
+                end
+            end
+        end
+        declarations[def.name] = true
+        return_protocol[def.name] = def
+    end
+    -- Detects a message or a enum that is used as a field type in message
+    -- but not defined in protocol. Allows a type be defined after usage
+    for def_type, declared in pairs(declarations) do
+        if not declared then
+            error(('Type %q is not declared'):format(def_type))
+        end
+    end
+    return setmetatable(return_protocol, protocol_mt)
+end
+
+local function is_number64(value)
+    if type(value) == 'cdata' and (ffi.istype(int64_t, value) or
+        ffi.istype(uint64_t, value)) then
+        return true
+    end
+end
+
+local function validate_length(value)
+    local MAX_LEN = 2^32
+    if string.len(value) > MAX_LEN then
+        error("Too long string to be encoded")
+    end
+end
+
+local function encode_len(field_id, value)
+    local wire_type = 2
+    return string.format('%s%s%s',
+        wireformat.encode_uint(bit.lshift(field_id, 3) + wire_type),
+        wireformat.encode_uint(string.len(value)),
+        value)
+end
+
+local encode_field
+
+local function code_rep_packed(protocol, name, field_name, field_id, data)
+    local buf = {}
+    protocol[name]['field_by_name'][field_name]['repeated'] = false
```

Comment:
This is no-no: it modifies a table that we don't own.

Reply:
Fixed, tested by multiple encoding using the same scheme